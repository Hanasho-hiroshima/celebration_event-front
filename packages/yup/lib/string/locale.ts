import { StringLocale } from 'yup/lib/locale'
import { Message } from 'yup/lib/types'

export const string: StringLocale & {
  katakana: StringLocale['email']
  furigana: Message
} = {
  length: ({ label, length }) =>
    (label ? label + 'は' : '') + `${length}文字で入力してください`,
  min: ({ label, min }) =>
    (label ? label + 'は' : '') + `${min}文字以上入力してください`,
  max: ({ label, max }) =>
    (label ? label + 'は' : '') + `${max}文字以下にしてください`,
  matches: ({ label, regex }) =>
    (label ? label + 'は' : '') + `以下の形式で入力してください: ${regex}`,
  email: ({ label }) =>
    (label ? label + 'は' : '') + `正しいメールアドレスを入力してください`,
  url: ({ label }) =>
    (label ? label + 'は' : '') + `正しいURLを入力してください`,
  uuid: ({ label }) =>
    (label ? label + 'は' : '') + `正しいUUIDを入力してください`,
  trim: ({ label }) =>
    (label ? label + 'は' : '') + `前後の空白を取り除いてください`,
  lowercase: ({ label }) =>
    (label ? label + 'は' : '') + `小文字のみ入力してください`,
  uppercase: ({ label }) =>
    (label ? label + 'は' : '') + `大文字のみ入力してください`,
  katakana: ({ label }) =>
    (label ? label + 'は' : '') + '全角カタカナのみ入力してください',
  furigana: ({ label }) =>
    (label ? label + 'は' : '') +
    '全角カタカナまたはひらがなで入力してください',
}
