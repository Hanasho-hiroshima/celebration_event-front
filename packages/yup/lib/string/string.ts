import { addMethod } from 'yup/lib/index'
import setLocale from 'yup/lib/setLocale'
import BaseSchema from 'yup/lib/schema'
import StringSchema, { create as stringCreate } from 'yup/lib/string'
import { AnyObject, Maybe, Message } from 'yup/lib/types'
import { string as stringLocale } from './locale'

export const isKatakana = (kana: string): boolean => {
  return /^([ァ-ン]|ヴ|ー| |　)*$/.test(kana || '')
}
export const isHiragana = (text: string): boolean => {
  return /^([ぁ-ん]|ゔ|ー| |　)*$/.test(text || '')
}

addMethod<StringSchema>(stringCreate, 'katakana', function () {
  return this.test('katakana', function (value, testContext) {
    if (isKatakana(value || '')) {
      return true
    }

    return testContext.createError({
      message: stringLocale.katakana,
    })
  })
})

addMethod<StringSchema>(
  stringCreate,
  'furigana',
  function (message: Message = stringLocale.furigana) {
    return this.test({
      message,
      name: 'furigana',
      test(value) {
        if (value == null) {
          return true
        }
        if (isKatakana(value) || isHiragana(value)) {
          return true
        }
        return false
      },
    })
  }
)

declare module 'yup' {
  // eslint-disable-next-line
  interface StringSchema<
    TType extends Maybe<string> = string | undefined,
    TContext extends AnyObject = AnyObject,
    TOut extends TType = TType
  > extends BaseSchema<TType, TContext, TOut> {
    katakana(): StringSchema<TType, TContext>
    furigana(message?: Message): StringSchema<TType, TContext>
  }
}

setLocale({
  string: stringLocale,
})

export default StringSchema

export { stringCreate }
