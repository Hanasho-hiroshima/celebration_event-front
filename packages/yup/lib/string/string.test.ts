import { AnySchema } from 'yup/lib/schema'
import ValidationError from 'yup/lib/ValidationError'
import { stringCreate } from './string'

const VALUES = ['TEXT', ' ', '　', 'テキスト', 'ﾃｷｽﾄ', null, undefined]
const label = 'フィールド'

const doTests = (
  schema: AnySchema,
  values: Array<string | null | undefined>
): void => {
  values.forEach((value) => {
    test(`${value}`, async () => {
      const result = await schema
        .validate(value, { strict: true, abortEarly: false })
        .then(() => true)
        .catch((err: ValidationError) => err.errors)
      expect(result).toMatchSnapshot()
    })
  })
}

describe('validation', () => {
  describe('katakana', () => {
    const schema = stringCreate().katakana()
    describe('no label', () => {
      doTests(schema, VALUES)
    })

    describe('with label', () => {
      const schemaWithLabel = schema.clone().label(label)
      doTests(schemaWithLabel, VALUES)
    })
  })

  describe('furigana', () => {
    const values = ['なまえ', 'ナマエ', 'ﾅﾏｴ', ' ', '　', null, undefined]
    const schema = stringCreate().furigana()
    describe('no label', () => {
      doTests(schema, values)
    })

    describe('with label', () => {
      const schemaWithLabel = schema.label(label)
      doTests(schemaWithLabel, values)
    })
  })
})
