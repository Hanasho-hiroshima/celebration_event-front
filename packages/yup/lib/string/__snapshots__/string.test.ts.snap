// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation furigana no label   1`] = `true`;

exports[`validation furigana no label null 1`] = `
[
  "this must be a \`string\` type, but the final value was: \`null\`.
 If "null" is intended as an empty value be sure to mark the schema as \`.nullable()\`",
]
`;

exports[`validation furigana no label undefined 1`] = `true`;

exports[`validation furigana no label 　 1`] = `true`;

exports[`validation furigana no label なまえ 1`] = `true`;

exports[`validation furigana no label ナマエ 1`] = `true`;

exports[`validation furigana no label ﾅﾏｴ 1`] = `
[
  "全角カタカナまたはひらがなで入力してください",
]
`;

exports[`validation furigana with label   1`] = `true`;

exports[`validation furigana with label null 1`] = `
[
  "フィールド must be a \`string\` type, but the final value was: \`null\`.
 If "null" is intended as an empty value be sure to mark the schema as \`.nullable()\`",
]
`;

exports[`validation furigana with label undefined 1`] = `true`;

exports[`validation furigana with label 　 1`] = `true`;

exports[`validation furigana with label なまえ 1`] = `true`;

exports[`validation furigana with label ナマエ 1`] = `true`;

exports[`validation furigana with label ﾅﾏｴ 1`] = `
[
  "フィールドは全角カタカナまたはひらがなで入力してください",
]
`;

exports[`validation katakana no label   1`] = `true`;

exports[`validation katakana no label TEXT 1`] = `
[
  "全角カタカナのみ入力してください",
]
`;

exports[`validation katakana no label null 1`] = `
[
  "this must be a \`string\` type, but the final value was: \`null\`.
 If "null" is intended as an empty value be sure to mark the schema as \`.nullable()\`",
]
`;

exports[`validation katakana no label undefined 1`] = `true`;

exports[`validation katakana no label 　 1`] = `true`;

exports[`validation katakana no label テキスト 1`] = `true`;

exports[`validation katakana no label ﾃｷｽﾄ 1`] = `
[
  "全角カタカナのみ入力してください",
]
`;

exports[`validation katakana with label   1`] = `true`;

exports[`validation katakana with label TEXT 1`] = `
[
  "フィールドは全角カタカナのみ入力してください",
]
`;

exports[`validation katakana with label null 1`] = `
[
  "フィールド must be a \`string\` type, but the final value was: \`null\`.
 If "null" is intended as an empty value be sure to mark the schema as \`.nullable()\`",
]
`;

exports[`validation katakana with label undefined 1`] = `true`;

exports[`validation katakana with label 　 1`] = `true`;

exports[`validation katakana with label テキスト 1`] = `true`;

exports[`validation katakana with label ﾃｷｽﾄ 1`] = `
[
  "フィールドは全角カタカナのみ入力してください",
]
`;
